pkg/agent/proxy/topology.go

```diff
@@ -115,21 +115,26 @@ func (p *proxier) categorizeEndpoints(endpoints map[string]k8sproxy.Endpoint, sv
 
 // canUseTopology returns true if topology aware routing is enabled and properly configured in this cluster. That is,
 // it checks that:
-// - The TopologyAwareHints feature is enabled.
-// - The "service.kubernetes.io/topology-aware-hints" annotation on this Service is set to "Auto".
-// - The node's labels include "topology.kubernetes.io/zone".
-// - All of the Endpoints for this Service have a topology hint.
-// - At least one Endpoint for this Service is hinted for this Node's zone.
+//   - The TopologyAwareHints or ServiceTrafficDistribution feature is enabled.
+//   - If ServiceTrafficDistribution feature is not enabled, then the "service.kubernetes.io/topology-aware-hints"
+//     annotation on this Service should be set to "Auto" or "auto".
+//   - The node's labels include "topology.kubernetes.io/zone".
+//   - All of the Endpoints for this Service have a topology hint.
+//   - At least one Endpoint for this Service is hinted for this Node's zone.
 func (p *proxier) canUseTopology(endpoints map[string]k8sproxy.Endpoint, svcInfo k8sproxy.ServicePort) bool {
-	if !p.topologyAwareHintsEnabled {
+	if !p.topologyAwareHintsEnabled && !p.serviceTrafficDistributionEnabled {
 		return false
 	}
-	hintsAnnotation := svcInfo.HintsAnnotation()
-	if hintsAnnotation != "Auto" && hintsAnnotation != "auto" {
-		if hintsAnnotation != "" && hintsAnnotation != "Disabled" && hintsAnnotation != "disabled" {
-			klog.InfoS("Skipping topology aware Endpoint filtering since Service has unexpected value", "annotationTopologyAwareHints", v1.DeprecatedAnnotationTopologyAwareHints, "hints", hintsAnnotation)
+
+	// Ignore the value of hintsAnnotation if the ServiceTrafficDistribution feature is enabled.
```

Comment:
Do you know why the code is slightly different for kube-proxy? https://github.com/kubernetes/kubernetes/blob/6fc0a69044f1ac4c13841ec4391224a2df241460/pkg/proxy/topology.go#L153-L161

Reply:
More generally, I wonder why this code is necessary for correctness. It seems to me that the real consumer of `service.kubernetes.io/topology-mode` and `.spec.trafficDistribution` is the EndpointSlice controller, and not the Service proxy. @tnqn do you know?

Reply:
Good catch! The current code is wrong. I made #6607 to fix it.

Reply:
@antoninbas both EndpointSlice controller and Service proxy are the consumers of the configuration.

- EndpointSlice controller adds hints to each endpoint when the feature is enabled.
- Service proxy consumes hints of endpoints when the feature is enabled, instead of unconditionally using the hints.

Reply:
Ah, I think I get your point. You are talking about each Service's configuration, instead of the global configuration. I agree with you this looks duplicate, and inconsistent for `service.kubernetes.io/topology-mode` and `.spec.trafficDistribution`. Perhaps it's just because the feature is implemented by different authors.

Reply:
Yes, I find the inconsistency strange and that's also related to my other comment. This seems to be the existing logic:
1) if both features are disabled, ignore the hints
2) if the TrafficDistribution feature is disabled (hence, the TopologyAwareHints feature is enabled), and we don't see an annotation, ignore the hints
3) otherwise, consume the hints

For consistency, IMO it should have been:
1) Consume the hints if and only if at least one the of the following conditions is true
  - the TrafficDistribution feature is enabled and `.spec.trafficDistribution` is set
  - the TopologyAwareHints feature is enabled and we see an annotation

However, I only see the current implementation being an issue if Feature Gates are set inconsistently between the proxy and the EndpointSlice controller. And we should probably just follow what kube-proxy is doing here.

Reply:
Since we have decided to keep the implementation consistent with kube-proxy, there is no need to consume `.spec.trafficDistribution` directly and we don't need to update K8s dependencies just yet.